### ISSUE TITLE
Problem with CHR and FONT LOAD

### ISSUE THREAD

**Issue by ldir-hector** [spotlessmind1975/ugbasic#570]
Previously I used FONT LOAD to redefine the font used for numbers.
Then I wanted to use it to redefine some other characters higher in the Ascii table, but without success.
I got Warning message, but then I realised that the warning "IMAGE converter unsupported for the given screen mode(2) was already present with the working example.

So I tried to see if my png file was wrong, and tried to use it with 
FONT LOAD "tank.png" TO 48
PRINT "0" and it worked !

So I did this for example : 

```
BITMAP ENABLE(160,200,16)
CLS

FONT LOAD "tank.png" TO 64

PRINT "-0";CHR(64)

FONT LOAD "tank.png" TO 96
PRINT "-0";CHR(96)

FONT LOAD "tank.png" TO 126
PRINT "-0";CHR(126)
```

And I got this : 
![image](https://user-images.githubusercontent.com/118670680/219958153-bc3ca217-647d-4e33-a616-6574a02132d2.png)

So it seems that if the Ascii number is to high, the FONT LOAD isn't working.

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector, and **thank you for your bug report!**

The bug was related to the fact that `FONT LOAD` works with screen codes while `CHR` works with ASCII code. These two encodings decode differently for some character sets, especially lowercase and uppercase characters. However, there was an error if the character exceeded the ASCII value of 120 (which is the last ASCII character that needs decoding).

Fixed on COLDFIX v1.13 [rev. 20230219.1310], see below:

![image](https://user-images.githubusercontent.com/11718606/220476758-711abad5-14dd-4a67-923d-87d25adf2223.png)

**Comment by ldir-hector** [comment]
Thanks for the bug fixing, but what do you mean by screen codes ?
And why do you say that that the last Ascii character that needs decoding is 120 ? 
I believed that Ascii characters were normalized on 7 bits for every one, with the first 32 that are commands characters and then with the remaining bit, each country did
what they want to introduce local characters.



----- Mail original -----
De: "Marco Spedaletti" ***@***.***>
À: "spotlessmind1975/ugbasic" ***@***.***>
Cc: "ldir-hector" ***@***.***>, "Mention" ***@***.***>
Envoyé: Mardi 21 Février 2023 23:51:29
Objet: Re: [spotlessmind1975/ugbasic] Problem with CHR and FONT LOAD (Issue #570)





Hello @ldir-hector , and thank you for your bug report! 

The bug was related to the fact that FONT LOAD works with screen codes while CHR works with ASCII code. These two encodings decode differently for some character sets, especially lowercase and uppercase characters. However, there was an error if the character exceeded the ASCII value of 120 (which is the last ASCII character that needs decoding). 

Fixed on COLDFIX v1.13 [rev. 20230219.1310], see below: 

image

**Comment by spotlessmind1975** [comment]
Hello @ldir-hector, I'm sorry for being too brief.
I'll try to explain better what I meant, especially in the context of ugBASIC.

Generally speaking,  8-bit ASCII encoding came after the introduction of 8-bit computers. Before that encoding, each manufacturer used its own proprietary encoding, also for lower 7-bit chars. To make matters worse, there were at least two encodings in every computer: the video processor encoding (each code also called "screen code") and the CPU encoding (each code also called "character code"). Some computers (such as ATARI) even had three encodings, and the additional one was keyboard encoding (each code also called "key code").

It follows that, when I was faced with the need to write a "simple" text on the screen, I had to lean towards using screen codes and not ASCII encoding, that has no meaning for the video processor (i.e. che letter 'A' is 65 for ASCII, 1 for SCREEN CODE). This is what the `PRINT` routine does, in the version for home computers that have a "text" video frame. The graphical version does the same, "simulating" the correspondence screen code - location in memory map of the font.

The most common transcoding is the one linked to the MOS VIC II processor, or rather it is the one that I found myself facing first. ATARI was the second, and so on. According to this encoding, there is a "character transposition" between this table and the ASCII table. In short, you have to do a series of subtractions or additions. Nothing transcendental. These are operations that must be carried out for ASCII codes lower than 120.

The same technique has been adopted on the Amstrad CPC 664, where in this home computer there is no "text" frame buffer and, therefore, no ROM/PLA that directly "maps" screen codes to characters (there is no such thing as "screen code"). It follows that the standard font being used, which is the same as the Commodore 64, requires the same decoding from the ASCII mapping since it has been composed following screen codes and not ASCII codes.

As for the first 32 ASCII codes, they are control characters according to late 80's terminal convention, but they are not for ugBASIC. To ensure the portability of strings from one processor to another, and from one computer to another, the meaning of the first codes was changed to implement "embedded" string commands.
